agents/template/src/main/scala/org/broadinstitute/transporter/kafka/TransferStream.scala

```diff
@@ -0,0 +1,105 @@
+package org.broadinstitute.transporter.kafka
+import java.nio.ByteBuffer
+
+import cats.effect.IO
+import cats.implicits._
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import io.circe._
+import io.circe.jawn.JawnParser
+import io.circe.syntax._
+import org.apache.kafka.streams.KafkaStreams
+import org.apache.kafka.streams.scala.StreamsBuilder
+import org.broadinstitute.transporter.queue.Queue
+import org.broadinstitute.transporter.transfer.{
+  TransferResult,
+  TransferRunner,
+  TransferStatus
+}
+
+/**
+  * Kafka stream defining how the "request" and "response" topics
+  * of a Transporter queue should be plugged together.
+  */
+object TransferStream {
+  import org.apache.kafka.streams.scala.Serdes._
+  import org.apache.kafka.streams.scala.ImplicitConversions._
+
+  /**
+    * Construct a stream which, when started, will pull requests
+    * from a Transporter queue, execute the requests using a runner,
+    * then push the results back to the Transporter manager.
+    *
+    * @param config configuration for the underlying Kafka components
+    *               running the stream
+    * @param queue Transporter queue describing the Kafka topics to
+    *              pull from / push into
+    * @param runner component which can actually perform data transfer
+    *
+    * @tparam In expected model for messages pulled from the "request"
+    *            topic of the given queue
+    */
+  def build[In: Decoder](
+    config: KStreamsConfig,
+    queue: Queue,
+    runner: TransferRunner[In]
+  ): IO[KafkaStreams] = {
+    val builder = new StreamsBuilder()
+    val jsonParser = new JawnParser()
+    val logger = Slf4jLogger.getLogger[IO]
+
+    /*
+     * Kafka Streams supports deserializers for higher-level types than `Array[Byte]`,
+     * but the interface for those serializers doesn't allow for transforming decoding
+     * errors into output messages.
+     *
+     * Since we rely on the agent reporting a status for every message it processes,
+     * we use the lowest-level deserializer at the stream entrypoint, then roll our own
+     * decoding / error handling logic.
+     */
+    builder
+      .stream[String, Array[Byte]](queue.requestTopic)
```

Comment:
if  a topic is requested it is presumably not available on the queue any more. couple questions which may seem obvious.
1) What prevents multiple agents picking up the same topic?
2) What ensures that if the agent working on a topic fails we don't lose that transfer? (not just errors I'm thinking more  the vm running the agent died or was shut down)